Add LPUART driver for the STM32G4 series

kywwilson11 opened the pull request:
## Summary
Add LPUART Driver for STM32G4:

- arch/arm/include/stm32/chip.h: define STM32_NLPUART

- arch/arm/src/stm32/Kconfig: Add options for configuring LPUART1. Left option out for LPUART 1-wire as it is not implemented for the STM32G4. 

- arch/arm/src/stm32/hardware/stm32g4xxr_pinmap.h (and legacy): Added additional definitions for LPUART1 CTS, RTS, RX and TX. 

- arch/arm/src/stm32/hardware/stm32g4xxxx_uart.h: Defined LPUART register addresses. USART offsets are compatible with LPUART and were reused here. Also added LPUART_CR1_ALLINTS as it differed slightly from UART_CR1_ALLINTS (fewer interrupt bits to clear). 

- arch/arm/src/stm32/stm32_lowputc.c: Defined STM32 console registers for LPUART1. Added code for setting BRR and PRESC LPUART registers. This code is selected only if CONSOLE_LPUART > 0 && defined(CONFIG_STM32_STM32G4XXX)

- arch/arm/src/stm32/stm32_serial.c: Many changes. Code was written to leave USART implementation as unchanged as possible. Added up_dev_s member islpuart to differentiate lpuarts from usarts. Added new implementation for calculating baud rate and prescaler for lpuarts. Added code to define DMA registers for STM32G4 LPUART. 

- arch/arm/src/stm32/stm32_uart.h: Added support for LPUART. Created new preprocessor definition CONSOLE_LPUART. 

- boards/arm/stm32/nucleo-g474re/configs/lpuartnsh/defconfig: Added new board config for the nucleo-g474re. This used the existing nsh config as a base, changed the console to LPUART1, and added configurations for dma1, dma2, dmamux1. The lpuart is the default serial port connected to the virtual com port on the nucleo-g474re. 

- boards/arm/stm32/nucleo-g474re/include/board.h: Added gpio definitions for lpuart1 tx, rx, cts, and rts. 

## Impact
Add support for LPUARTs in the arch/arm/\*/stm32 family of boards. This should only impact the STM32G4 series as none of the other chips defined in arch/arm/\*/stm32 have LPUART peripherals. Code was written to change as little as possible of the existing UART implementation. 

## Testing
nucleo-g474re:lpuartnsh (for LPUART testing)
nucleo-g474re:nsh (for USART3 testing)

LPUART Testing:
- Baud Rate: Tested baud rates from 300 baud to 250,000 baud.
- Parity: Tested none, odd, and even
- Stop Bits: Tested 1 and 2
- DMA: Tested using the LPUART1 console. Demonstrated basic functionality with TXDMA, RXDMA, and both RXDMA/TXDMA enabled. 

USART Testing: To make sure USART functionality was not broken, tested basic functionality using USART3 as a console. Tried baud rates 9600 and 115200, along with various parity and stop bit settings.

nuttxpr commented:
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



## NuttX PR Requirements Review

**Overall:** The PR appears to mostly meet the NuttX requirements. The summary and impact sections are well-detailed and provide a good overview of the changes and their implications. 

**Suggestions for Improvement:**

* **Testing:**
    * While the testing description is helpful, including actual testing logs (even if brief snippets) would significantly strengthen the PR. 
    * Consider adding tests to the NuttX regression test suite (`nuttx/testing`) to ensure ongoing functionality of the LPUART driver. 
    * Specify the exact versions of tools used for building and testing (e.g., "GCC 11.3.0").

* **Documentation:**
    * The impact section mentions documentation updates may be required.  
        * If documentation updates are needed, ensure these are included in the PR.  
        * If no updates are needed, explicitly state that within the "Impact on documentation" section.

* **Compatibility:**
    * Briefly elaborate on backward compatibility. Does this change affect existing configurations that were using the same pins or resources as the LPUART?

**Minor Notes:**

*  Typos: 
    * "descibe" -> "describe"
    * "arch(s) / board(s) / driver(s)" -> Consistent wording would be "architectures / boards / drivers"

**Addressing these points will enhance the PR's quality and clarity, making it easier for reviewers to evaluate and merge.**

kywwilson11 commented:
Update:
1. added "ifdef CONFIG_STM32_LPUART1" statements around lpuart related code that was causing build issues for non-stm32g4 builds. 
2. Changed the lpuart BRR calculation to use a 64-bit integer cast. This implementation was copied from the stm32l4 serial driver. This results in a slightly more accurate baud rate.

lupyuen commented on arch/arm/src/stm32/stm32_serial.c:
```diff
@@ -1452,52 +1552,142 @@ static void up_set_format(struct uart_dev_s *dev)
+       * Sum the upper and lower calculations to get the BRR
+       */
+
+      brr = (((uint64_t)apbclock_whole << 8) + (priv->baud >> 1)) / priv->baud;
```
Please break the long line thanks!

lupyuen commented on arch/arm/src/stm32/stm32_serial.c:
```diff
@@ -1785,8 +1981,11 @@ static int up_setup(struct uart_dev_s *dev)
-  regval &= ~(USART_CR2_STOP_MASK | USART_CR2_CLKEN | USART_CR2_CPOL |
-              USART_CR2_CPHA | USART_CR2_LBCL | USART_CR2_LBDIE);
+  if (priv->islpuart == true)
+      regval &= ~(USART_CR2_STOP_MASK | USART_CR2_CLKEN);
```
Let's wrap this line with braces: `{ regval @= ... }`

lupyuen commented on arch/arm/src/stm32/stm32_serial.c:
```diff
@@ -1785,8 +1981,11 @@ static int up_setup(struct uart_dev_s *dev)
+  if (priv->islpuart == true)
+      regval &= ~(USART_CR2_STOP_MASK | USART_CR2_CLKEN);
+  else
+      regval &= ~(USART_CR2_STOP_MASK | USART_CR2_CLKEN | USART_CR2_CPOL |
```
Let's wrap this line with braces: `{ regval @= ... }`

lupyuen commented:
We have an error in the Config File: https://github.com/apache/nuttx/actions/runs/11182711564/job/31107864104
```text
  Normalize nucleo-g474re/lpuartnsh
25a26
> CONFIG_LPUART1_SERIAL_CONSOLE=y
37a39,41
> CONFIG_STM32_DMA1=y
> CONFIG_STM32_DMA2=y
> CONFIG_STM32_DMAMUX1=y
41d44
< CONFIG_LPUART1_SERIAL_CONSOLE=y
44,46d46
< CONFIG_STM32_DMA1=y
< CONFIG_STM32_DMA2=y
< CONFIG_STM32_DMAMUX1=y
```
I think the sequence of Config Options might be incorrect. Try running `tools/refresh.sh` [(explained here)](https://lupyuen.github.io/articles/pr#appendix-nuttx-configuration-files)
```bash
$ cd nuttx
$ tools/refresh.sh --silent nucleo-g474re:lpuartnsh
## Remember to commit the updated `defconfig`!
```

kywwilson11 commented:
I made a few commits this morning. When I initially broke up the brr calculation into two lines, I ended up truncating the result into 32-bits, then dividing, which gave an incorrect value. I fixed this and committed, but then forgot to remove a TAB. I then committed and pushed this final update. Everything should be good to go. Thank you!

lupyuen commented:
Yep thank you so much for contributing the LPUART Driver :-)
